src/utils/word-api.ts

```diff
@@ -0,0 +1,205 @@
+/* eslint-disable no-console */
+import { CitationResult, StatefulCitation } from "citeproc";
+
+class WordApi {
+  static async insertEmptyContentControl(): Promise<unknown> {
+    return Word.run((context: Word.RequestContext) => {
+      const getSelection = context.document.getSelection();
+      const contentControl = getSelection.insertContentControl();
+      contentControl.tag = "JABREF-CITATION-NEW";
+      contentControl.appearance = "BoundingBox";
+      return context.sync();
+    }).catch((error) => {
+      console.log(`Error: ${JSON.stringify(error)}`);
+      if (error instanceof OfficeExtension.Error) {
+        console.log(`Debug info: ${JSON.stringify(error.debugInfo)}`);
+      }
+    });
+  }
+
+  async setCitations(
+    data: Array<CitationResult>,
+    citationByIndex: Array<StatefulCitation>
+  ): Promise<unknown> {
+    return Word.run(async (context: Word.RequestContext) => {
+      const citationContentControls = context.document.contentControls;
+      context.load(citationContentControls, "tag, length");
+      return context.sync().then(() => {
+        data.forEach((res: CitationResult) => {
+          const position = res[0];
+          const tag = this.getCitationTagForPosition(citationByIndex, position);
```

Comment:
Something I don't understand here: why don't you produce the tag directly from the citation `res` but have this additional step with the position? Possibly for this you would need to put the information that is currently in "citationByIndex[position]" directly in the citation object, but I think this would be cleaner. But probably I'm missing some constraints why you need to have the citationByIndex list around.

Reply:
When we process citation data(by calling register citation), processor returns updated `citationByIndex`(adds citation data of the processed citation in citationByindex) and `citationData` aka `res`. The structure of `res` is `[position, citationText, CitationID]`. In this res array, the CitationID is the Id assigned by the processor to this particular citation. That's about the structure of res.
Now, let's discuss about the citationByIndex: 

## State of Document
### Empty state
When we start with a new document with no citation in it, We initialize our addin with an empty `citationByIndex` array which we pass with `initProcessor`. Now processor is initialized with no citation registered in it. Let's call it **State 0**. Now you start working on your document, inserting citations in it. Now for some reason, you closed the document with citations in. Let's call it **state 1** when you closed the document.

### Initializing processor to a previously saved state.
Now you start working on the doc again. And you initialize the add-in, which also initializes the processor. But now we have citations in the document, which we have to also pass to the processor with `initProcessor`. We do so by regenerating citationByIndex. We iterate over all of the citations, get their tag according to their respective position, and add these tags in citationByIndex. Now we have an array with citation data for all positions. We pass this with initProcessor and initialize the processor to the previously saved state, aka **state 1**. Processor returns rebuild data which we use to update the text of the citations if there is any modification done in the database. 

All reasoning why we are using the citationByIndex array object to produce the tag is we can reuse this data the initialize the processor to the current state of the document. 

Now the question is, why do we need to initialize the processor to a particular state?
If we initialize the processor to a particular state, we can interact(edit/update/modify) with previously saved citations.
Another reason is the numeric-based citation style. The processor can use citationByIndex and assign the numbers correctly.

## Updating citationByIndex
We don't manually interact with the data of citationByIndex array's object. 
### When we update or construct citationByIndex ?
When we initialize addin we construct citationByIndex. Another scenario when we update citation by index is when we insert citation. This update is just to check if any previously cited reference is deleted or copy/pasted. Just to sync processor's citationByIndex with citesupport citationByindex.

Reply:
Thanks a lot for the explanation. That makes it very clear why the citationByIndex structure is needed. I suggest you put this text almost verbatim in the citation worker file.

It sounds like the citationByIndex is somewhat of an implementation detail. In particular, the only word-specific thing is that one needs to build the citationByIndex when the word document is loaded (i.e the getCitationByIndex method further below). However, here there is no real need to use the citationByIndex directly if I understand you correctly. Thus I would propose to extend the data object by essentially what is in `citationByIndex[position]`. I.e data would look like an array of objects
```ts
{
   position: number
   citationText: string
   citation: StatefulCitation
}
```
So you essentially push the lookup position -> citation to a level higher. Not sure how far you can push it but I noticed that at least all callers of citesupport.setCitations also get the citationByIndex.

What do you think?